### PR TITLE
Added JS Unit Test With Mocha and JS Zip Strategies

### PR DESCRIPTION
**Pull request by Innos**

### CODE REVIEW

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsPreprocessExecuteAndCheckExecutionStrategy.cs [hunk 1]
```diff
@@ -71,13 +72,14 @@ public class NodeJsPreprocessExecuteAndCheckExecutionStrategy : ExecutionStrateg
...
-var content = ''";
+var content = '';
+var adapterFunction = " + AdapterFunctionPlaceholder;
 
         protected virtual string JsCodePostevaulationCode => string.Empty;
 
         protected virtual string JsCodeEvaluation => @"
     var inputData = content.trim().split(EOL);
-    var result = code.run(inputData);
+    var result = adapterFunction(inputData,code.run);
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsPreprocessExecuteAndCheckExecutionStrategy.cs]: space after the comma

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsPreprocessExecuteAndRunCodeAgainstUnitTestsWithMochaExecutionStrategy.cs [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+            string jsdomModulePath,
+            string jqueryModulePath,
+            string handlebarsModulePath,
+            string sinonModulePath,
+            string sinonChaiModulePath,
+            string underscoreModulePath,
+            int baseTimeUsed,
+            int baseMemoryUsed)
+            : base(
+                nodeJsExecutablePath, mochaModulePath, chaiModulePath, jsdomModulePath, jqueryModulePath,
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsPreprocessExecuteAndRunCodeAgainstUnitTestsWithMochaExecutionStrategy.cs]: each param on separate line

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+            string sinonModulePath,
+            string sinonChaiModulePath,
+            string underscoreModulePath,
+            string browserifyModulePath,
+            string babelifyModulePath,
+            string ecmaScriptImportPluginPath,
+            int baseTimeUsed,
+            int baseMemoryUsed)
+            : base(
+                nodeJsExecutablePath, mochaModulePath, chaiModulePath, jsdomModulePath, jqueryModulePath,
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs]: each param on separate line

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+
+    public class NodeJsZipPreprocessExecuteAndRunUnitTestsWithDomAndMochaExecutionStrategy :
+        IoJsPreprocessExecuteAndRunJsDomUnitTestsExecutionStrategy
+    {
+        protected const string AppJsFileName = "app.js";
+
+        protected const string SubmissionFileName = "_$submission";
+
+        public NodeJsZipPreprocessExecuteAndRunUnitTestsWithDomAndMochaExecutionStrategy
+            (
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs]: `(` must be on the line above

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+            int baseTimeUsed,
+            int baseMemoryUsed)
+            : base(
+                nodeJsExecutablePath, mochaModulePath, chaiModulePath, jsdomModulePath, jqueryModulePath,
+                handlebarsModulePath, sinonModulePath, sinonChaiModulePath, underscoreModulePath, baseTimeUsed,
+                baseMemoryUsed)
+        {
+            if (!Directory.Exists(browserifyModulePath))
+            {
+                throw new ArgumentException($"Browsrify not found in: {browserifyModulePath}",
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs]: `$"Browsrify not found in: {browserifyModulePath}",` -> new line

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+                    nameof(browserifyModulePath));
+            }
+
+            if (!Directory.Exists(babelifyModulePath))
+            {
+                throw new ArgumentException($"Babel not found in: {babelifyModulePath}", nameof(babelifyModulePath));
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs]: `$"Babel not found in: {babelifyModulePath}",` -> new line

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+            }
+
+            if (!Directory.Exists(ecmaScriptImportPluginPath))
+            {
+                throw new ArgumentException($"ECMAScript2015ImportPluginPath not found in: {ecmaScriptImportPluginPath}",
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs]: `$"ECMAScript2015ImportPluginPath not found in: {ecmaScriptImportPluginPath}",` -> new line

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+delete clearInterval;
+delete setImmediate;
+delete clearImmediate;
+delete module;
+delete require;
+delete msg;
+
+process.exit = function () {};";
+
+        protected override string JsCodeTemplate => this.JsCodeRequiredModules +
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs]: `this.JsCodeRequiredModules +` must be on the next line, indented with **one** tab.

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+        }";
+
+        protected override string JsCodePostevaulationCode => @"
+        });
+    });
+});";
+
+    public override ExecutionResult Execute(ExecutionContext executionContext)
+        {
+            ExecutionResult result = new ExecutionResult();
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs]: Use `var`

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+
+            result.IsCompiledSuccessfully = true;
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs]: you can skip this line if you use *object initializer* -> `var result = new ExecutionResult { IsCompiledSuccessfully = true }; `

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+
+            // Copy and unzip the file (save file to WorkingDirectory)
+            string submissionFilePath = this.CreateSubmissionFile(executionContext);
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs]: Use `var`

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+            this.ProgramEntryPath = this.FindProgramEntryPath();
+
+            // Replace the placeholders in the JS Template with the real values
+            string codeToExecute = this.PreprocessJsSubmission(this.JsCodeTemplate, this.ProgramEntryPath, executionContext.TaskSkeletonAsString);
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs]: Use `var`

Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs]: Please try not to go beyond 120 symbols in a row. -> move params of `this.PreprocessJsSubmission` at a new lines

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+
+            // Save code to file
+            var codeSavePath = FileHelpers.SaveStringToTempFile(codeToExecute);
+
+            // Create a Restricted Process Executor
+            IExecutor executor = new RestrictedProcessExecutor();
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs]: Use `var`

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+
+            // Create a Checker using the information from the Execution Context
+            IChecker checker = Checker.CreateChecker(executionContext.CheckerAssemblyName, executionContext.CheckerTypeName, executionContext.CheckerParameter);
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs]: Use `var`

Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs]: Please try not to go beyond 120 symbols in a row. -> move params of `Checker.CreateChecker` at new lines

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs (hunk 11 above)
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs]: If you don't need `submissionFilePath`, you can remove it.

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+        {
+            using (var zipFile = ZipFile.Read(fileToUnzip))
+            {
+                foreach (var entry in zipFile)
+                {
+                    entry.Extract(outputDirectory, ExtractExistingFileAction.OverwriteSilently);
+                }
+            }
+
+            System.IO.File.Delete(fileToUnzip);
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs]: There is no need to specify namespace of the File class.

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs [hunk 16]
```diff
@@ -0,0 +1,322 @@
...
+        }
+
+        protected virtual string FindProgramEntryPath()
+        {
+            var files = new List<string>(Directory.GetFiles(this.WorkingDirectory, "app.js", SearchOption.AllDirectories));
+            if (files.Count == 0)
+            {
+                throw new ArgumentException($"'{AppJsFileName}' file not found in output directory!", nameof(this.WorkingDirectory));
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs]: Please try not to go beyond 120 symbols in a row. -> move params of `ArgumentException` at new lines

File: Open Judge System/Workers/OJS.Workers.LocalWorker/Batch Files/InstallService.bat [hunk 17]
```diff
@@ -1,6 +1,6 @@
 NET STOP "OJS Local Worker Service"
 sc delete "OJS Local Worker Service"
-timeout 10
+timeout 5
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.LocalWorker/Batch Files/InstallService.bat]: Why?

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsPreprocessExecuteAndRunCodeAgainstUnitTestsWithMochaExecutionStrategy.cs [hunk 18]
```diff
@@ -0,0 +1,255 @@
...
+                        }).forEach(function (prop) {
+                            global[prop] = window[prop];
+                        });
+                    done();
+                    }
+            });
+        });
+
+        describe('Tests',function(){
+
```
Comment (YanaSlavcheva) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsPreprocessExecuteAndRunCodeAgainstUnitTestsWithMochaExecutionStrategy.cs]: no blank line here

**Pull request merged**